src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,469 @@
+
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+#define CUPTI_CALL(call)                                                       \
+do {                                                                           \
+    CUptiResult _status = call;                                                \
+    if (_status != CUPTI_SUCCESS) {                                            \
+        const char *errstr;                                                    \
+        cuptiGetResultString(_status, &errstr);                                \
+        spdlog::error("function {} failed with error {}", #call, errstr);      \
+        exit(EXIT_FAILURE);                                                    \
+    }                                                                          \
+} while (0)
+
+#define PTHREAD_CALL(call)                                                         \
+do {                                                                               \
+    int _status = call;                                                            \
+    if (_status != 0) {                                                            \
+        spdlog::error("function {} failed with error code {}", #call, _status);    \
+        exit(EXIT_FAILURE);                                                        \
+    }                                                                              \
+} while (0)
+
+typedef enum {
+    FI_TRAP,
+    FI_ASSERT,
+    FI_RETURN_VALUE
+} FaultInjectionType;
+
+typedef struct {
+    volatile uint32_t initialized;
+    CUpti_SubscriberHandle  subscriber;
+
+    bool dynamic;
+    int terminateThread;
+    pthread_t dynamicThread;
+    pthread_rwlock_t configLock = PTHREAD_RWLOCK_INITIALIZER;
+
+    boost::property_tree::ptree configRoot;
+    boost::optional<boost::property_tree::ptree&> driverFaultConfigs = boost::none;
+    boost::optional<boost::property_tree::ptree&> runtimeFaultConfigs = boost::none;
+} injGlobalControl;
+injGlobalControl globalControl;
+
+// Function Declarations
+static void atExitHandler(void);
+
+static void CUPTIAPI faultInjectionCallbackHandler(void *userdata, CUpti_CallbackDomain domain, CUpti_CallbackId cbid, void *cbInfo);
+
+extern int STDCALL InitializeInjection(void);
+
+
+#if defined(__cplusplus)
+}
+#endif
+
+static CUptiResult cuptiInitialize(void);
+static void readFaultInjectorConfig(void);
+static void parseConfig(boost::property_tree::ptree const& pTree);
+static void *dynamicReconfig(void *args);
+
+// Function Definitions
+
+static void
+globalControlInit(void) {
+    spdlog::trace("globalControlInit of fault injection");
+    globalControl.initialized = 0;
+    globalControl.subscriber = 0;
+    globalControl.terminateThread = 0;
+    readFaultInjectorConfig();
+}
+
+static void
+registerAtExitHandler(void) {
+    // Register atExitHandler
+    atexit(&atExitHandler);
+}
+
+
+static void
+atExitHandler(void) {
```

Comment:
Curious if this is being pedantic, or is there something that is being cleaned up here that normal OS process teardown will not automatically handle?

Reply:
Actually, I should read this more closely. I'm not sure if the usage model is:
- tag a separately running Spark RAPIDS process to inject a fault, or,
- inject fault from within the same Spark process.

Reply:
> dynamically linked to the process under test

Ah, then @jlowe might have a point about CUPTI cleanup likely happening on process exit.
That said, my vote would be to keep the `atExitHandler()` anyway.